In [1]:
import pandas as pd

df = pd.read_csv("../data/processed/fraud_processed.csv")
df.head()

,purchase_value,source,browser,sex,age,class,country,time_since_signup,hour_of_day,day_of_week,user_transaction_count,device_transaction_count
0,47,SEO,Safari,F,30,0,Australia,3632215.0,3,6,1,1
1,15,SEO,Chrome,F,34,0,Thailand,1231637.0,20,2,1,1
2,44,Ads,IE,M,29,0,China,1997534.0,23,5,1,2
3,55,Direct,Chrome,F,30,0,China,7640897.0,16,5,1,1
4,51,SEO,Chrome,F,37,0,Thailand,10249579.0,4,1,1,1


In [2]:
X = df.drop("class", axis=1)
y = df["class"]

In [3]:
X = pd.get_dummies(X, columns=["source", "browser", "sex", "country"], drop_first=True)

In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

In [5]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=42)

X_train_res, y_train_res = smote.fit_resample(X_train, y_train)

ModuleNotFoundError: No module named 'imblearn'

In [6]:
from imblearn.over_sampling import SMOTE

ModuleNotFoundError: No module named 'imblearn'

In [7]:
import sys
sys.executable

'c:\\Users\\HP\\AppData\\Local\\Programs\\Python\\Python311\\python.exe'

In [1]:
import sys
sys.executable

'c:\\Users\\HP\\AppData\\Local\\Programs\\Python\\Python311\\python.exe'

In [1]:
import sys
sys.executable

'c:\\Users\\HP\\AppData\\Local\\Programs\\Python\\Python311\\python.exe'

In [1]:
import sys
sys.executable

'c:\\Users\\HP\\AppData\\Local\\Programs\\Python\\Python311\\python.exe'

In [1]:
import sys
print(sys.executable)

c:\Users\HP\AppData\Local\Programs\Python\Python311\python.exe


In [1]:
import pandas as pd

fraud_df = pd.read_csv("../data/processed/fraud_processed.csv")
credit_df = pd.read_csv("../data/processed/creditcard_processed.csv")

FileNotFoundError: [Errno 2] No such file or directory: '../data/processed/creditcard_processed.csv'

In [2]:
import os

print(os.listdir("../data"))

['processed', 'raw']


In [3]:
print(os.listdir("../data/processed"))

['fraud_processed.csv']


In [4]:
import pandas as pd

fraud_df = pd.read_csv("../data/processed/fraud_processed.csv")

fraud_df.head()

,purchase_value,source,browser,sex,age,class,country,time_since_signup,hour_of_day,day_of_week,user_transaction_count,device_transaction_count
0,47,SEO,Safari,F,30,0,Australia,3632215.0,3,6,1,1
1,15,SEO,Chrome,F,34,0,Thailand,1231637.0,20,2,1,1
2,44,Ads,IE,M,29,0,China,1997534.0,23,5,1,2
3,55,Direct,Chrome,F,30,0,China,7640897.0,16,5,1,1
4,51,SEO,Chrome,F,37,0,Thailand,10249579.0,4,1,1,1


In [5]:
print(fraud_df.shape)

(129146, 12)


In [6]:
print(fraud_df.columns.tolist())

['purchase_value', 'source', 'browser', 'sex', 'age', 'class', 'country', 'time_since_signup', 'hour_of_day', 'day_of_week', 'user_transaction_count', 'device_transaction_count']


In [7]:
import os

os.listdir("../data/raw")

['creditcard.csv', 'Fraud_Data.csv', 'IpAddress_to_Country.csv']

In [8]:
import pandas as pd

credit_df = pd.read_csv("../data/raw/creditcard.csv")

print(credit_df.shape)

credit_df.head()

(284807, 31)


,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0
1,0.0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0
2,1.0,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0
3,1.0,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0
4,2.0,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0


In [ ]:
fraud_df.info()

In [ ]:
fraud_df.info()

In [1]:
fraud_df.info()

NameError: name 'fraud_df' is not defined

In [2]:
import pandas as pd

fraud_df = pd.read_csv("../data/processed/fraud_processed.csv")

fraud_df.head()

,purchase_value,source,browser,sex,age,class,country,time_since_signup,hour_of_day,day_of_week,user_transaction_count,device_transaction_count
0,47,SEO,Safari,F,30,0,Australia,3632215.0,3,6,1,1
1,15,SEO,Chrome,F,34,0,Thailand,1231637.0,20,2,1,1
2,44,Ads,IE,M,29,0,China,1997534.0,23,5,1,2
3,55,Direct,Chrome,F,30,0,China,7640897.0,16,5,1,1
4,51,SEO,Chrome,F,37,0,Thailand,10249579.0,4,1,1,1


In [3]:
fraud_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 129146 entries, 0 to 129145
Data columns (total 12 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   purchase_value            129146 non-null  int64  
 1   source                    129146 non-null  str    
 2   browser                   129146 non-null  str    
 3   sex                       129146 non-null  str    
 4   age                       129146 non-null  int64  
 5   class                     129146 non-null  int64  
 6   country                   129146 non-null  str    
 7   time_since_signup         129146 non-null  float64
 8   hour_of_day               129146 non-null  int64  
 9   day_of_week               129146 non-null  int64  
 10  user_transaction_count    129146 non-null  int64  
 11  device_transaction_count  129146 non-null  int64  
dtypes: float64(1), int64(7), str(4)
memory usage: 11.8 MB


In [4]:
import os

os.listdir("../data/raw")

['creditcard.csv', 'Fraud_Data.csv', 'IpAddress_to_Country.csv']

In [5]:
import pandas as pd

fraud_df = pd.read_csv("../data/processed/fraud_processed.csv")

print("Shape:", fraud_df.shape)
fraud_df.head()

Shape: (129146, 12)


,purchase_value,source,browser,sex,age,class,country,time_since_signup,hour_of_day,day_of_week,user_transaction_count,device_transaction_count
0,47,SEO,Safari,F,30,0,Australia,3632215.0,3,6,1,1
1,15,SEO,Chrome,F,34,0,Thailand,1231637.0,20,2,1,1
2,44,Ads,IE,M,29,0,China,1997534.0,23,5,1,2
3,55,Direct,Chrome,F,30,0,China,7640897.0,16,5,1,1
4,51,SEO,Chrome,F,37,0,Thailand,10249579.0,4,1,1,1


In [6]:
X = fraud_df.drop("class", axis=1)
y = fraud_df["class"]

print(X.shape)
print(y.shape)

(129146, 11)
(129146,)


In [1]:
X = pd.get_dummies(
    X,
    columns=["source", "browser", "sex", "country"],
    drop_first=True
)

print(X.shape)

NameError: name 'pd' is not defined

In [2]:
import pandas as pd

fraud_df = pd.read_csv("../data/processed/fraud_processed.csv")

print(fraud_df.shape)

(129146, 12)


In [3]:
X = fraud_df.drop("class", axis=1)
y = fraud_df["class"]

print(X.shape)
print(y.shape)

(129146, 11)
(129146,)


In [4]:
X = pd.get_dummies(
    X,
    columns=["source", "browser", "sex", "country"],
    drop_first=True
)

print(X.shape)

(129146, 194)


In [5]:
print(X.shape)

(129146, 194)


In [6]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Training set:", X_train.shape)
print("Test set:", X_test.shape)

Training set: (103316, 194)
Test set: (25830, 194)


In [7]:
print("Class counts:")
print(y_train.value_counts())

print("\nClass percentages:")
print(y_train.value_counts(normalize=True) * 100)

Class counts:
class
0    93502
1     9814
Name: count, dtype: int64

Class percentages:
class
0    90.500987
1     9.499013
Name: proportion, dtype: float64


In [8]:
!pip install imbalanced-learn

In [9]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=42)

X_train_smote, y_train_smote = smote.fit_resample(
    X_train,
    y_train
)

print(y_train_smote.value_counts())

class
0    93502
1    93502
Name: count, dtype: int64


In [10]:
print(y_train_smote.value_counts(normalize=True) * 100)

class
0    50.0
1    50.0
Name: proportion, dtype: float64


In [11]:
print(y_train.value_counts())
print(y_train.value_counts(normalize=True) * 100)

class
0    93502
1     9814
Name: count, dtype: int64
class
0    90.500987
1     9.499013
Name: proportion, dtype: float64


In [12]:
import nbformat

nb = nbformat.read("../notebooks/modeling.ipynb", as_version=4)

for i, cell in enumerate(nb.cells):
    print(f"\n{'='*50}")
    print(f"CELL {i+1} ({cell.cell_type})")
    print(f"{'='*50}")

    print(cell.source[:500])


CELL 1 (code)
import pandas as pd

df = pd.read_csv("../data/processed/fraud_processed.csv")
df.head()

CELL 2 (code)
X = df.drop("class", axis=1)
y = df["class"]

CELL 3 (code)
X = pd.get_dummies(X, columns=["source", "browser", "sex", "country"], drop_first=True)

CELL 4 (code)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

CELL 5 (code)
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=42)

X_train_res, y_train_res = smote.fit_resample(X_train, y_train)

CELL 6 (code)
from imblearn.over_sampling import SMOTE

CELL 7 (code)
import sys
sys.executable

CELL 8 (code)
import sys
sys.executable

CELL 9 (code)
import sys
sys.executable

CELL 10 (code)
import sys
sys.executable

CELL 11 (code)
import sys
print(sys.executable)

CELL 12 (code)
import pandas as pd

fraud_df = pd.read_csv("../data/processed/fraud_processed.csv")
credit_df = pd.read_

In [13]:
import nbformat

nb = nbformat.read("../notebooks/modeling.ipynb", as_version=4)

print("Total cells:", len(nb.cells))

Total cells: 40


In [14]:
import nbformat

nb = nbformat.read("../notebooks/modeling.ipynb", as_version=4)

for i, cell in enumerate(nb.cells):
    first_line = cell.source.split("\n")[0] if cell.source else "EMPTY CELL"
    print(f"Cell {i+1}: {cell.cell_type} -> {first_line}")

Cell 1: code -> import pandas as pd
Cell 2: code -> X = df.drop("class", axis=1)
Cell 3: code -> X = pd.get_dummies(X, columns=["source", "browser", "sex", "country"], drop_first=True)
Cell 4: code -> from sklearn.model_selection import train_test_split
Cell 5: code -> from imblearn.over_sampling import SMOTE
Cell 6: code -> from imblearn.over_sampling import SMOTE
Cell 7: code -> import sys
Cell 8: code -> import sys
Cell 9: code -> import sys
Cell 10: code -> import sys
Cell 11: code -> import sys
Cell 12: code -> import pandas as pd
Cell 13: code -> import os
Cell 14: code -> print(os.listdir("../data/processed"))
Cell 15: code -> import pandas as pd
Cell 16: code -> print(fraud_df.shape)
Cell 17: code -> print(fraud_df.columns.tolist())
Cell 18: code -> import os
Cell 19: code -> import pandas as pd
Cell 20: code -> fraud_df.info()
Cell 21: code -> fraud_df.info()
Cell 22: code -> fraud_df.info()
Cell 23: code -> import pandas as pd
Cell 24: code -> fraud_df.info()
Cell 25: code ->

# Task 2: Model Building and Evaluation

In [1]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import (
    confusion_matrix,
    classification_report,
    f1_score,
    average_precision_score
)

In [2]:
lr = LogisticRegression(
    max_iter=1000,
    random_state=42
)

lr.fit(X_train_smote, y_train_smote)

NameError: name 'X_train_smote' is not defined

In [3]:
print(X_train.shape)

NameError: name 'X_train' is not defined

In [1]:
import sys
print(sys.executable)

c:\Users\HP\fraud-detection\.venv\Scripts\python.exe


In [2]:
import pandas as pd

df = pd.read_csv("../data/processed/fraud_processed.csv")

In [3]:
X = df.drop("class", axis=1)
y = df["class"]

In [4]:
X = pd.get_dummies(
    X,
    columns=["source", "browser", "sex", "country"],
    drop_first=True
)

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [6]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=42)

X_train_smote, y_train_smote = smote.fit_resample(
    X_train,
    y_train
)

In [7]:
print(X_train.shape)
print(X_train_smote.shape)

(103316, 194)
(187004, 194)


In [8]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(
    max_iter=1000,
    random_state=42
)

lr.fit(X_train_smote, y_train_smote)

print("Logistic Regression trained successfully!")

Logistic Regression trained successfully!


In [9]:
y_pred_lr = lr.predict(X_test)

y_pred_proba_lr = lr.predict_proba(X_test)[:, 1]

In [10]:
from sklearn.metrics import (
    confusion_matrix,
    classification_report,
    f1_score,
    average_precision_score
)

print("F1 Score:")
print(f1_score(y_test, y_pred_lr))

print("\nAUC-PR:")
print(average_precision_score(y_test, y_pred_proba_lr))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_lr))

print("\nClassification Report:")
print(classification_report(y_test, y_pred_lr))

F1 Score:
0.4186431329288472

AUC-PR:
0.6039395212134672

Confusion Matrix:
[[20096  3280]
 [  936  1518]]

Classification Report:
              precision    recall  f1-score   support

           0       0.96      0.86      0.91     23376
           1       0.32      0.62      0.42      2454

    accuracy                           0.84     25830
   macro avg       0.64      0.74      0.66     25830
weighted avg       0.89      0.84      0.86     25830



### Logistic Regression Baseline Model

A Logistic Regression classifier was trained as an interpretable baseline model using the fraud transaction dataset.

The model was trained on data balanced using SMOTE after applying a stratified train-test split. Evaluation was conducted on the untouched test set using metrics appropriate for imbalanced classification problems.

#### Results

| Metric   | Score |
| -------- | ----- |
| F1 Score | 0.419 |
| AUC-PR   | 0.604 |
| Accuracy | 0.84  |

#### Interpretation

The model achieved a recall of 0.62 for fraudulent transactions, meaning that it successfully detected approximately 62% of fraud cases. However, the precision of 0.32 indicates that a significant number of legitimate transactions were incorrectly flagged as fraudulent.

This behavior is common in fraud detection systems where detecting fraudulent activity is prioritized over minimizing false alarms.

The baseline model provides a useful benchmark for comparison with more powerful ensemble methods such as Random Forest and XGBoost.


In [11]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    f1_score,
    average_precision_score,
    confusion_matrix,
    classification_report
)

rf = RandomForestClassifier(
    n_estimators=100,
    max_depth=10,
    random_state=42,
    n_jobs=-1
)

rf.fit(X_train_smote, y_train_smote)

rf_pred = rf.predict(X_test)
rf_prob = rf.predict_proba(X_test)[:, 1]

print("Random Forest trained successfully!")

print("\nF1 Score:")
print(f1_score(y_test, rf_pred))

print("\nAUC-PR:")
print(average_precision_score(y_test, rf_prob))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, rf_pred))

print("\nClassification Report:")
print(classification_report(y_test, rf_pred))

Random Forest trained successfully!

F1 Score:
0.6926429120738272

AUC-PR:
0.6505173577668943

Confusion Matrix:
[[23280    96]
 [ 1103  1351]]

Classification Report:
              precision    recall  f1-score   support

           0       0.95      1.00      0.97     23376
           1       0.93      0.55      0.69      2454

    accuracy                           0.95     25830
   macro avg       0.94      0.77      0.83     25830
weighted avg       0.95      0.95      0.95     25830



# Model Comparison and Selection

Two classification models were evaluated for fraud detection:

1. Logistic Regression (Baseline Model)
2. Random Forest (Ensemble Model)

## Performance Comparison

| Metric | Logistic Regression | Random Forest |
|----------|----------|----------|
| F1 Score | 0.419 | 0.693 |
| AUC-PR | 0.604 | 0.651 |

## Discussion

The Random Forest model outperformed Logistic Regression across the primary evaluation metrics.

Although Logistic Regression achieved higher recall for the fraud class, it generated a large number of false positives, which could negatively impact customer experience by incorrectly blocking legitimate transactions.

Random Forest achieved a much higher F1-score and AUC-PR while maintaining excellent precision, indicating a better balance between fraud detection and false alarm reduction.

## Selected Model

Random Forest was selected as the final model because it provided the best overall performance on the highly imbalanced fraud detection dataset.

In [12]:
from sklearn.model_selection import StratifiedKFold, cross_val_score

cv = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

f1_scores = cross_val_score(
    rf,
    X_train_smote,
    y_train_smote,
    cv=cv,
    scoring="f1"
)

print("F1 Scores:")
print(f1_scores)

print("\nMean F1:")
print(f1_scores.mean())

print("\nStandard Deviation:")
print(f1_scores.std())

F1 Scores:
[0.8389133  0.83802098 0.84040417 0.82679258 0.82794591]

Mean F1:
0.8344153864925176

Standard Deviation:
0.0058147791638427015


# Cross-Validation Results

To obtain a more reliable estimate of model performance, 5-fold Stratified Cross-Validation was applied to the Random Forest model.

## Results

Mean F1 Score: 0.8344

Standard Deviation: 0.0058

## Interpretation

The Random Forest model demonstrated stable performance across all folds, with a low standard deviation indicating consistent behavior on different subsets of the data.

These results suggest that the model generalizes well and is unlikely to be overfitting the training data.

## Model Comparison

| Model | F1 Score | AUC-PR |
|---------|---------|---------|
| Logistic Regression | 0.419 | 0.604 |
| Random Forest | 0.693 | 0.651 |

### Selected Model

Random Forest was selected as the final model because it achieved higher F1 Score and AUC-PR compared to Logistic Regression.

The model also demonstrated stable performance during 5-fold Stratified Cross Validation with a mean F1 Score of 0.834 and a standard deviation of 0.0058.

These results indicate that Random Forest is better suited for detecting fraudulent transactions while maintaining a balance between precision and recall.

In [13]:
!pip install shap

  Using cached cloudpickle-3.1.2-py3-none-any.whl.metadata (7.1 kB)
   ---------------------------------------- 0.0/554.9 kB ? eta -:--:--
   ---------------------------------------- 0.0/554.9 kB ? eta -:--:--
   ------------------ --------------------- 262.1/554.9 kB ? eta -:--:--
   ------------------ --------------------- 262.1/554.9 kB ? eta -:--:--
   ---------------------------------------- 554.9/554.9 kB 669.9 kB/s  0:00:00
Using cached cloudpickle-3.1.2-py3-none-any.whl (22 kB)
   ---------------------------------------- 0.0/38.1 MB ? eta -:--:--
   ---------------------------------------- 0.0/38.1 MB ? eta -:--:--
   ---------------------------------------- 0.3/38.1 MB ? eta -:--:--
    --------------------------------------- 0.5/38.1 MB 882.6 kB/s eta 0:00:43
    --------------------------------------- 0.8/38.1 MB 958.5 kB/s eta 0:00:39
    --------------------------------------- 0.8/38.1 MB 958.5 kB/s eta 0:00:39
   - -------------------------------------- 1.0/38.1 MB 1.0 MB